cpu/native/periph/spi.c

```diff
@@ -0,0 +1,284 @@
+                    cs
+                    );
+                assert(false);
+            }
+            DEBUG("spi_init: %d:%d %s (fd 0x%x)\n", bus, cs,
+                  conf->device_filename[cs], fd);
+            state->fd[cs] = fd;
+        }
+        else {
+            DEBUG("spi_init: %d:%d Unused\n", bus, cs);
+        }
+    }
+    DEBUG("spi_init: done\n");
+}
+
+int spi_init_cs(spi_t bus, spi_cs_t cs)
+{
+    if (bus >= SPI_NUMOF) {
+        return SPI_NODEV;
+    }
+    else if (cs != SPI_CS_UNDEF && cs >= SPI_MAXCS) {
+        return SPI_NOCS;
+    }
+    else if (device_state[bus].fd[cs] < 0) {
+        return SPI_NOCS;
+    }
+    return SPI_OK;
+}
+
+void spi_init_pins(spi_t bus)
+{
+    (void)bus;
+    /* Nothing to do here, as the kernel driver does the pin management */
+}
+
+void spi_release(spi_t bus)
+{
+    DEBUG("spi_release(%d)\n", bus);
+    if (bus < SPI_NUMOF) {
+        mutex_unlock(&(device_state[bus].lock));
+    }
+}
+
+static int spi_set_params(int fd, bool hwcs, spi_mode_t mode, spi_clk_t clk)
+{
+    uint8_t spi_mode = mode | (hwcs ? 0 : SPI_NO_CS);
+
+    if (real_ioctl(fd, SPI_IOC_WR_MODE, &spi_mode) < 0) {
+        return SPI_NOMODE;
+    }
+    uint32_t ioctl_clk = clk;
+    if (real_ioctl(fd, SPI_IOC_WR_MAX_SPEED_HZ, &ioctl_clk) < 0) {
+        return SPI_NOCLK;
+    }
+    return SPI_OK;
+}
+
+void spi_transfer_bytes(spi_t bus, spi_cs_t cs, bool cont,
+                        const void *out, void *in, size_t len)
+{
+    if (bus < SPI_NUMOF && (cs == SPI_CS_UNDEF || cs < SPI_MAXCS)) {
+        int fd = (cs == SPI_CS_UNDEF) ?
+                 spidev_get_first_fd(&(device_state[bus])) :
+                 device_state[bus].fd[cs];
+
+        if (fd >= 0) {
+            struct spi_ioc_transfer spi_tf;
+            memset(&spi_tf, 0, sizeof(spi_tf));
+
+            intptr_t out_addr = (intptr_t)out;
+            intptr_t in_addr = (intptr_t)in;
+
+            /* Leaving speed_hz as zero uses the value from spi_acquire */
+            spi_tf.bits_per_word = 8;
+            /*
+             * The kernel documentation is a bit ambiguous about how to use the
+             * cs_change value ("True to deselect device"). It seems like
+             * setting it to true leaves the CS line actually low (=selected)
+             * after transmission.
+             */
+            spi_tf.cs_change = cont;
+            spi_tf.len = len;
+            spi_tf.rx_buf = (uint64_t)in_addr;
+            spi_tf.tx_buf = (uint64_t)out_addr;
+
+            if (ENABLE_DEBUG) {
+                DEBUG("spi_transfer_bytes: -> ");
+                for (size_t i = 0; i < len; i++) {
+                    DEBUG(" %02x", out != NULL ? ((uint8_t *)out)[i] : 0);
+                }
+                DEBUG("\n");
+            }
+
+            if (real_ioctl(fd, SPI_IOC_MESSAGE(1), &spi_tf) < 0) {
+                DEBUG("spi_transfer_bytes: ioctl failed\n");
+            }
+            else {
+                if (ENABLE_DEBUG) {
```

Comment:
Are those elaborate DEBUG functions really useful?
I'd remove them for the sake of readability.

Reply:
Printing outgoing and incoming data was quite helpful during development, but one might reduce it to `ioctl failed` or `transferred %d bytes`.

Reply:
I can imagine :)
But it looks better with the scaffolding removed once the implementation is working.